File: modules/phono-phonegap-ios/libphono/src/iosaudio/SpeexCodec.m
/*
 * Copyright 2011 Voxeo Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#import "SpeexCodec.h"
#include <speex/speex_callbacks.h>


@implementation SpeexCodec

- (id) init{
    return [self initWide:NO];
}

- (id) initWide:(BOOL)wide{
    [super init];
    wideBand = wide;
    if (wide) {
        decoder_st = speex_decoder_init(&speex_wb_mode);
        encoder_st = speex_encoder_init(&speex_wb_mode);
        int tmp=0;
        speex_encoder_ctl(encoder_st, SPEEX_SET_VBR, &tmp);
        tmp=5;
        speex_encoder_ctl(encoder_st, SPEEX_SET_QUALITY, &tmp);
        tmp=2;
        speex_encoder_ctl(encoder_st, SPEEX_SET_COMPLEXITY, &tmp);
        tmp = 16000;
        speex_encoder_ctl(encoder_st,SPEEX_SET_SAMPLING_RATE,&tmp);  
    } else {
        decoder_st = speex_decoder_init(&speex_nb_mode);
        encoder_st = speex_encoder_init(&speex_nb_mode);
        int tmp=0;
        speex_encoder_ctl(encoder_st, SPEEX_SET_VBR, &tmp);
        tmp=5;
        speex_encoder_ctl(encoder_st, SPEEX_SET_QUALITY, &tmp);
        tmp=1;
        speex_encoder_ctl(encoder_st, SPEEX_SET_COMPLEXITY, &tmp);
        tmp = 8000;
        speex_encoder_ctl(encoder_st,SPEEX_SET_SAMPLING_RATE,&tmp);
    }
	speex_bits_init(&eBits);
	speex_bits_init(&dBits);
    
    
    return self;
}

- (NSString *) getName{
    return @"SPEEX";
}
- (NSInteger) getRate{
    return wideBand ? 16000:8000;
}

- (int) ptype{
    return wideBand?103:102;
}
- (BOOL) decode:(NSData *)wireData audioData:(NSMutableData *)audioData{
    int len = wideBand?640:320;
    [audioData setLength:len];
    uint8_t * wire = (uint8_t *) [wireData bytes];
    int16_t *audio = (int16_t *) [audioData mutableBytes];
    speex_bits_reset(&dBits);
    speex_bits_read_from(&dBits, (char *) wire,[wireData length]);
    speex_decode_int(decoder_st, &dBits, audio);
    return YES;
}


- (BOOL) encode:(NSData *)audioData wireData:(NSMutableData *)wireData{
    [wireData setLength:160];
    uint8_t * wire = (uint8_t *) [wireData mutableBytes];
    int16_t *audio = (int16_t *) [audioData bytes];    
    speex_bits_reset(&eBits);
    speex_encode_int(encoder_st, audio, &eBits);
    int nbBytes = speex_bits_write(&eBits, (char*) wire, 160);
    [wireData setLength:nbBytes];

    return YES;
}
- (void)dealloc {
    speex_encoder_destroy(encoder_st);
    speex_decoder_destroy(decoder_st);
    encoder_st = nil;
    decoder_st = nil;
    [super dealloc];
}
@end

File: modules/phono-java-audio/src/java/com/phono/codecs/speex/SpeexCodec.java
/*
 * Copyright 2011 Voxeo Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.phono.codecs.speex;

import com.phono.audio.codec.CodecFace;
import com.phono.audio.codec.DecoderFace;
import com.phono.audio.codec.EncoderFace;
import com.phono.srtplight.Log;
import java.io.StreamCorruptedException;
import org.xiph.speex.SpeexDecoder;
import org.xiph.speex.SpeexEncoder;

public class SpeexCodec implements CodecFace, EncoderFace, DecoderFace {

    SpeexEncoder _spxe;
    SpeexDecoder _spxd;
    int _sampleRate;
    long _iaxcn;
    String _name;
    int _aframesz;
    int _speexmode;

    public SpeexCodec(boolean wide) {
        _speexmode = wide ? 1:0;
        _sampleRate = wide ? 16000 : 8000;
        _iaxcn = wide ? CodecFace.AUDIO_SPEEX16 : CodecFace.AUDIO_SPEEX;
        _name = wide ? "SPEEX" : "SPEEX";
        _spxe = new SpeexEncoder();
        int q = 9;
        _spxe.init(_speexmode, q, _sampleRate, 1);// _mode, _quality, _sampleRate, _channels);
        _spxe.getEncoder().setComplexity(5);
        Log.debug("Speex mode "+ _spxe.getEncoder().getMode()
                +" encoder configured with rate ="+_spxe.getEncoder().getSamplingRate()
                +" Quality ="+q
                +" compexity = "+_spxe.getEncoder().getComplexity()
                +" packet size = "+_spxe.getEncoder().getEncodedFrameSize()/8
                +" bitrate = "+_spxe.getEncoder().getBitRate()/1024.0);
        _spxd = new SpeexDecoder();
        _spxd.init(_speexmode, _sampleRate, 1, false);// _mode, _sampleRate, _channels, false);
        _aframesz = wide ? 320 : 160; // number of shorts in an audio frame;
    }


    public int getFrameSize() {
        return -1; // we don't know - it is officailly a vbr codec.
    }

    public int getFrameInterval() {
        return 20;
    }

    public long getCodec() {
        return _iaxcn;
    }

    public DecoderFace getDecoder() {
        return this;
    }

    public EncoderFace getEncoder() {
        return this;
    }

    public String getName() {
        return _name;
    }

    public float getSampleRate() {
        return _sampleRate;
    }

    public byte[] encode_frame(short[] audio) {
        _spxe.processData(audio, 0, audio.length);
        int sz = _spxe.getProcessedDataByteSize();
        byte[] wireOut = new byte[sz];
        int got = _spxe.getProcessedData(wireOut, 0);
        return wireOut;
    }

    public short[] decode_frame(byte[] bytes) {
        try {
            _spxd.processData(bytes, 0, bytes.length);
        } catch (StreamCorruptedException ex) {
            Log.error("Speex Decoder error " + ex.getMessage());
        }
        short audioOut[] = new short[_aframesz];
        int decsize = _spxd.getProcessedData(audioOut, 0);
        return audioOut;
    }

    public byte[] lost_frame(byte[] bytes, byte[] bytes1) {
        // only gives us a _decoded_ frame - we don;t know what to do with that....
        // todo....
        return null;
    }
}
